### ISSUE TITLE
Union inside dict values fails when it shouldn't

### ISSUE THREAD

**Issue by marcelgerber** [microsoft/pyright#4986]
## Describe the bug
I got a `SomeSet = set[T] | frozenset[T]` type alias.
When I now specify a dict of type `dict[str, SomeSet[T]]`, I would expect that I can use both `frozenset[int]` and `set[int]` as dict values. However, `set` works, whereas `frozenset` doesn't work.

## To Reproduce
Consider the following code:
```py
from typing import TypeVar

T = TypeVar("T")

SomeSet = set[T] | frozenset[T]


def do_something_with_set(some_dict: dict[str, SomeSet[T]]) -> list[T]:
    return []

do_something_with_set({"foo": set({1, 2, 3})}) # <-- this line works
do_something_with_set({"foo": frozenset({1, 2, 3})}) # <-- this line fails
```

You can also [check this out in Pyright playground.](https://pyright-playground.decorator-factory.su/?gzip=H4sIALY5QmQC_4WPzQ6CQAyE7_sUE7xAAgf1RtSXgHghhBDZlY3Akt0agsq7W_w7Ge2h6UynX1JlTQsae90dodveWEI69nJfWiFSbN_C91IvECIxrUwkse8kZWmOG5Q1F9k9pRCikgqVKRwHqWZoMWiqC177s1VU-kAx5p45siFeQL7NA0Q7NNrNIhbgspLOtkPG3O_Iq6eM8WI85mWIVYj1FEwBFthEETjtmNhJDMae3B_I55HfKFXqxt0Bk3EJsTYBAAA%3D)

The error message I get out of this is:
```
Argument of type "dict[str, frozenset[int]]" cannot be assigned to parameter "some_dict" of type "dict[str, SomeSet[T@do_something_with_set]]" in function "do_something_with_set"
  Type "frozenset[int]" cannot be assigned to type "SomeSet[T@do_something_with_set]"
    "frozenset[int]" is incompatible with "set[T@do_something_with_set]"
    "frozenset[int]" is incompatible with "frozenset[T@do_something_with_set]"
      TypeVar "_T_co@frozenset" is covariant
        Type "int" cannot be assigned to type "T@do_something_with_set"
```

## Expected behavior
This code should not throw any type errors.

## VS Code extension or command-line
I tried this both with:
- VS Code Pylance, `v2023.4.30`
- Command line pyright, `1.1.304`

## Additional context

I tried some more stuff, and it got even weirder then:
```py
from typing import Mapping, TypeVar

T = TypeVar("T")

SomeSet = set[T] | frozenset[T]
SomeSet2 = frozenset[T] | set[T]


def do_something_with_set_1(some_dict: dict[str, SomeSet[T]]) -> list[T]:
    return []

def do_something_with_set_2(some_dict: dict[str, SomeSet2[T]]) -> list[T]: # uses SomeSet2, where set and frozenset are swapped
    return []

def do_something_with_set_3(some_dict: Mapping[str, SomeSet[T]]) -> list[T]: # uses Mapping rather than dict
    return []


do_something_with_set_1({"foo": set({1, 2, 3})}) # <-- works
do_something_with_set_1({"foo": frozenset({1, 2, 3})}) # <-- fails

do_something_with_set_2({"foo": set({1, 2, 3})}) # <-- fails
do_something_with_set_2({"foo": frozenset({1, 2, 3})}) # <-- works

do_something_with_set_3({"foo": set({1, 2, 3})}) # <-- works
do_something_with_set_3({"foo": frozenset({1, 2, 3})}) # <-- works

```
[Pyright playground](https://pyright-playground.decorator-factory.su/?gzip=H4sIAIQ7QmQC_6VQwU6EMBS89ysmeIEEDgs3ov6BpyVeCCGNFGlcKGm7Ieu6_-6rwLpRYTX20OS9N2_ezNRatbCHXnbPkG2vtMUD710ZIjv04pFrxjLczYXvZV7A2Fa1Yiss9Y2weVbgDbVWr6Iby3keE-CyT7AJwFglalSqNIS0Dd0rB2mbksblxnfNspJPNoX7c2N1iImTtosA0T120rgiZaCnhd3rDnmxRhyvEsffmHGDvRHmDAgxNEIL5wG8qz6tgbvuQMGJ6vdykks5U-rrVmdBExiaWxIE2_Duw8_X22wp4KNXK-Wlzol_3IQga8kpOAV04DaKMCj9Yq4un-3_RFFzuTNLAuJrAsbta8urAkYPbCn6fySQ_EXAOzllnfVhAwAA)

This extended example shows that:
- Types `set[T] | frozenset[T]` and `frozenset[T] | set[T]` are different, which is unexpected to me
- `dict[K, V]` and `Mapping[K, V]` behave differently in this case.

I also [tested this same code out in mypy playground](https://mypy-play.net/?mypy=latest&python=3.11&gist=423367ebfe4c1c34b3aecd2def803b00), where it's running fine.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.305, which I just published. It will also be included in a future release of pylance.

**Comment by marcelgerber** [comment]
Cool, thank you very much for the quick turnaround!
